Pattern-matching compilation

gasche opened the pull request:
This is a PR that implements compilation of nested patterns, by the usual technique of matrix decomposition.

(I thought of many ways to do it efficiently at the bytecode level, and I can tell you about the stack-layout scheme I devised, but in the end I decided to implement lowering precisely to do it at a more source-like level, for portability to a different backend for example.)

It appears that this conflict with changes done in trunk since my last pull, I will fix those this evening.

I have not implemented or-patterns yet, but they are not too hard to add on top of that, and used in my miniml. This is the next step.

The input grammar for pattern sorts of sucks (children have to be simple patterns so you have to write `Foo((Bar z),(Foobar x)))` instead of `Foo(Bar z, Foobar x)`). Generalizing it requires adding precedence levels to avoid conflicts, so some reengineering of the pattern grammar. I also plan to do this, but later.

gasche commented on miniml/compiler/compile.scm:
```diff
@@ -1079,41 +1079,172 @@
+; instead of cons-pairs, as further specialized representations
+; are used during the transformation at higher arities
+; (3 for non-empty rows, 5 for non-empty rows with beheaded simple first pattern)
+; and cons-pairs quickly become inconvenient at higher arities.
```
I am planning to change this to use records instead of fixed-size lists (fixed-size lists look fine but they are used nowhere except in AST structures) at some point.

gasche commented:
I fixed the conflicts with the current master branch.

Ekdohibs reviewed (commented):
Thanks for the pull request! Most of the things that I think should be done are already in TODOs in the code, but I added some comments. Also, I think now that the matching uses variables for each intermediate value, we could simplify `'LSwitch` to act on a variable instead of a general expression, and avoid the need for pushing that variable to access its fields.

gasche commented on miniml/compiler/compile.scm:
```diff
@@ -1093,35 +1117,173 @@
+(define (simplify-row env arg row)
+  (match-let (((p ps e) row))
+   (match p
+    (('PVar v)
+     (let* ((p (list 'PAs (list 'PWild) v)))
+     (simplify-row env arg (list p ps e))))
+    (('PAs p v)
+     (let* ((e (list 'LLet v (lid->evar arg) e)))
```
Yes, but do we care? I would be curious to know if we can in fact see a performance overhead to naive compilation approaches (indeed the proposed compiler generates worse code than the current one for the shallow fragment).

As you suggest we could optimize variable-variable rebind. We could also try to not bind variables for subvalues that are unused. (In the OCaml compiler this is done with a "simplification" pass on the Lambda code that erases unused "alias" bindings and inlines the ones used exactly once; here it seems easier to deal with variable-variable rebind on the fly during bytecode production.)

gasche replied:
I compiled the `master` version of `hello.ml` with various versions of the compiler, measuring the size (in bytes) of the produced bytecode executable.

- 22_262 (22Kio): this PR
- 21_102 (21Kio): this PR plus support for variable rebinding
- 19_918 (20Kio): master

One way to read these numbers is that adding support for variable rebinding halves the bytecode-size overhead of this PR for shallow patterns. I am going to add this support as a further commit to the PR.

gasche replied:
Another fun number: 21_102 bytes is when all variable rebindings in the program are optimized away. If we optimize only those that are generated by the PVar/PAlias case of pattern-matching, we get a 21_782-bytes output. So when we think of this optimization as shaving the pattern-matching overhead from 2344 bytes to 1184 bytes (a 1160 bytes saving), it seems that we only save 480 bytes from the pattern-matching code, the rest coming from secondary benefits from this optimization (possibly other inefficiences introduced by the lowering of other constructs).

gasche commented on miniml/compiler/compile.scm:
```diff
@@ -1093,35 +1117,173 @@
+                     ; in the wildcard case, we add the row to the default group
+                     ; (representing any value whose head is not among the other groups)
+                     ; *and* to each group (their values also match a wildcard)
+                     ; this is why we need to have a pre-computed list of all groups
```
Yes indeed. I think that duplicating code is problematic in principle, so I was planning to tackle this when introducing or-patterns (which also duplicate the actions of some rows). During compilation we can detect which actions (= right-hand-side expressions) get duplicated. Then the question is what code to generate for those. There are three potential approaches:
1. We could simply generate local functions and jump to them; this gives fairly bad code, but we probably don't care so much (especially if we disable deduplication when the duplicated action is a variable or a constant).
2. We could manually generate bytecode labels and use branches; the problem is that, in the or-pattern case, the duplicated branches may have bound the same variables in different places in the stack, so to do this correctly the easier way is to re-push the shared variables in some agreed-upon order before branching. This sounds too painful to me.
3. We could implement support for OCaml's `Lstaticcach/Lstaticraise`, which are block-local exceptions (with parameters), used in particular for deduplication within the pattern-matching compiler; or we could expose this as non-leaking local functions that can only be tail-called. This gives a nice/principled structure to do (2).

gasche replied:
Note: in [Data.value_compare](https://github.com/Ekdohibs/camlboot/blob/74bdd14/interpreter/data.ml#L281) there is actually no issue, as the "default" actions are very small (either `failwith <string literal>` or `assert false`).

gasche replied:
I started looking at implementing Lstaticraise, or rather something like

```
let-exits
  j1 x y z = ...
  j2 x y = ...
in
  ....  exit j2 foo bar ...
```

but I am tempted to do some refactoring before that (ideally to separate a "typing environment" from a "compilation environment", with variables in both sides, one just for the shape and the other just for the location), and I would prefer to do this on top of the current state of the branch, rather than on top of master, because some things are cleaner in the branch, notably the handling of `"_"`.

I hesitate between:
- suggesting to decide that duplication is not a concern for now, and that this merged could be merged as-is (when other issues have been addressed, etc.), so that I can work on top of a better master to implement exits
- doing a more direct implementation of exits (yet another parameter to `compile-low-expr` that we would pass all over the place) for now

I am going to "yield" this work now and move to other things, so in any case I will probably not write more code today.

Note: it is not as easy as I thought it would be to detect which actions are duplicated. The issue is that the PR currently mixes matrix decomposition and code generation (each decomposition of the matrix compute submatrices, generates their code, and then generates the code for the matrix by composing them under a `Switch`), so by the time the information comes back to the caller of how actions are used, the code is already generated. Two ideas:
1. I could give up on the idea of only using a duplication-avoidance mechanism only for actions that are indeed duplicated, and just use it for any "big enough" action.
2. I could reify the submatrix decomposition into a datastructure (where actions are represented by symbolic names / indices), get the duplication information from that, and then re-traverse it to generate the code.

I am thinking of going for (2), which is probably not that much work.

gasche replied:
Note: I did implement the first half of (2), now matrix-decomposition returns a "matching tree", which does not touch the action expressions themselves, and is lowered into code in a second step.

gasche commented:
> we could simplify LSwitch to act on a variable instead of a general expression, and avoid the need for pushing that variable to access its fields.

I tried to implement this in https://github.com/gasche/camlboot/commit/e77c833f53ca16b801fec63b890840ae4b21ef38 , but I'm getting segfaults which I am too impatient to debug. (It's already a pain to debug the placement of parentheses in my new code before seeing whether it works, and then the silly type errors :sweat:) I think that the code-quality wins are fairly minor: we avoid a couple PUSH instructions, but then the code we generate to access the block is less amenable to instruction-compression than the `ACC0` of the current approach. There could be gains due to the fact that the compilation of default clauses becomes a mere rebind, but I decided to give up before seeing if that was the case.

Ekdohibs commented:
I found the problem with your switching-or-variable commit, and left a comment there. Still, it remains an open question about whether it would have benefits or not.

gasche commented:
I am currently looking at completing the "lowering" separation to make it nicer to implement let-exits -- so I won't look at switch compilation right away.

gasche commented:
I rebased on top of the current master (in preparation for rebasing on top of let-exits and using those). The testing changes mean that the code-size impact of the current PR is very easy to assess: the first commit increases bytecode sizes somewhat, and then the second commit (the rebinding optimization) reduces them back to a size close to the original size.

gasche commented:
I rebased this PR on top of #31 and #32, implemented deduplication of duplicated actions, and generalized the nested-pattern grammars so that they can be written in a natural style. In terms of feature set, I consider that this is sufficient for a first pattern-matching-compilation PR.

There is still a small code-size increase with this PR, but it is smaller than with the previous iteration. For example, the first PR had:

```
+++ b/miniml/compiler/test/labels.info.reference
-Bytecode size:    4374 bytes
+Bytecode size:    4406 bytes
```

while this PR has

```
+++ b/miniml/compiler/test/labels.info.reference
-Bytecode size:    3982 bytes
+Bytecode size:    3998 bytes
```

The work on action deduplication does not really reduce bytecode size in common cases, where there are low amounts of duplication for small right-hand-sides, so the deduplication saving does not compensate the extra code size of the exit handlers. I made sure that patterns.ml contains an action that is both duplicated and large, and there the savings are of course significant. In general this is not so much about the good-case behavior than about avoiding bad-case behaviors.

The next step is to implement or-patterns :cactus:

gasche commented:
As a warmup exercise, I implemented support for as-patterns `p as v`.

gasche commented:
I just pushed support for or-patterns. This completes the work I had in mind for pattern-matching compilation.

(The compilation of or-patterns is naive: the right-hand-side is not duplicated, but nested or-patterns will result in a fair amount of duplication of control-flow code by generating identical submatrices whose code could be shared. This is a large part of what, for example, "Compiling pattern-matching to good decision trees", Luc Maranget, 2008, is about. I don't think that this would be an issue in the interpreter codebase, so I am not planning to work on this.)

Ekdohibs commented:
Thanks a lot! Again, will merge once rebased. I think with this pull request, we will have most, in not all of the code needed to finally bootstrap ocamllex and to compile the interpreter.

gasche commented:
This is rebased.

(I recently implemented support for `module type S = ...` definitions, which are just ignored. I pushed this direction to the master branch.)

Ekdohibs commented:
Thanks !
